crates/meilisearch/src/routes/chat.rs

```diff
@@ -0,0 +1,466 @@
+
+        let choice = &mut response.choices[0];
+        match choice.finish_reason {
+            Some(FinishReason::ToolCalls) => {
+                let tool_calls = mem::take(&mut choice.message.tool_calls).unwrap_or_default();
+
+                let (meili_calls, other_calls): (Vec<_>, Vec<_>) =
+                    tool_calls.into_iter().partition(|call| call.function.name == "searchInIndex");
+
+                chat_completion.messages.push(
+                    ChatCompletionRequestAssistantMessageArgs::default()
+                        .tool_calls(meili_calls.clone())
+                        .build()
+                        .unwrap()
+                        .into(),
+                );
+
+                for call in meili_calls {
+                    let SearchInIndexParameters { index_uid, q } =
+                        serde_json::from_str(&call.function.arguments).unwrap();
+
+                    let (_, text) =
+                        process_search_request(&index_scheduler, &search_queue, index_uid, q)
+                            .await?;
+
+                    chat_completion.messages.push(ChatCompletionRequestMessage::Tool(
+                        ChatCompletionRequestToolMessage {
+                            tool_call_id: call.id,
+                            content: ChatCompletionRequestToolMessageContent::Text(text),
+                        },
+                    ));
+                }
+
+                // Let the client call other tools by themselves
+                if !other_calls.is_empty() {
+                    response.choices[0].message.tool_calls = Some(other_calls);
+                    break;
+                }
+            }
+            _ => break,
+        }
+    }
+
+    Ok(HttpResponse::Ok().json(response))
+}
+
+async fn streamed_chat(
+    index_scheduler: GuardedData<ActionPolicy<{ actions::CHAT_GET }>, Data<IndexScheduler>>,
+    search_queue: web::Data<SearchQueue>,
+    mut chat_completion: CreateChatCompletionRequest,
+) -> impl Responder {
+    let chat_settings = match index_scheduler.chat_settings().unwrap() {
+        Some(value) => serde_json::from_value(value).unwrap(),
+        None => ChatSettings::default(),
+    };
+
+    let mut config = OpenAIConfig::default();
+    if let Some(api_key) = chat_settings.api_key.as_ref() {
+        config = config.with_api_key(api_key);
+    }
+    if let Some(base_api) = chat_settings.base_api.as_ref() {
+        config = config.with_api_base(base_api);
+    }
+
+    setup_search_tool(&mut chat_completion, &chat_settings.prompts);
+
+    let (tx, rx) = tokio::sync::mpsc::channel(10);
+    let _join_handle = Handle::current().spawn(async move {
+        let client = Client::with_config(config.clone());
+        let mut global_tool_calls = HashMap::<u32, Call>::new();
+        let mut finish_reason = None;
+
+        'main: while finish_reason.map_or(true, |fr| fr == FinishReason::ToolCalls) {
```

Comment:
I am not sure we should run a new conversation when the original stream stopped without any finish reason. We should stop.
```suggestion
        'main: while finish_reason.map_or(false, |fr| fr == FinishReason::ToolCalls) {
```